core/src/tools/bscrypto.cc

```diff
@@ -28,166 +28,217 @@
  */
 
 #include "include/bareos.h"
+#include "lib/cli.h"
 #include "lib/crypto_cache.h"
 #include "lib/crypto_wrap.h"
 #include "lib/passphrase.h"
 #include "lib/scsi_crypto.h"
 
-static void usage()
+/**
+ * Read the key bits from the keyfile.
+ * - == stdin
+ */
+static void ReadKeyBits(const std::string& keyfile, char* data)
 {
-  fprintf(
-      stderr,
-      "\n"
-      "Usage: bscrypto <options> [<device_name>]\n"
-      "       -b              Perform base64 encoding of keydata\n"
-      "       -c              Clear encryption key\n"
-      "       -D <cachefile>  Dump content of given cachefile\n"
-      "       -d <nn>         Set debug level to <nn>\n"
-      "       -e              Show drive encryption status\n"
-      "       -g <keyfile>    Generate new encryption passphrase in keyfile\n"
-      "       -k <keyfile>    Show content of keyfile\n"
-      "       -p <cachefile>  Populate given cachefile with crypto keys\n"
-      "       -r <cachefile>  Reset expiry time for entries of given "
-      "cachefile\n"
-      "       -s <keyfile>    Set encryption key loaded from keyfile\n"
-      "       -v              Show volume encryption status\n"
-      "       -w <keyfile>    Wrap/Unwrap the key using RFC3394 aes-(un)wrap\n"
-      "                       using the key in keyfile as a Key Encryption "
-      "Key\n"
-      "       -?              print this message.\n"
-      "\n"
-      "       A keyfile named - is either stdin or stdout depending on the "
-      "option\n"
-      "\n");
+  int kfd = 0;
+  if (bstrcmp(keyfile.c_str(), "-")) {
+    kfd = 0;
+    fprintf(stdout, _("Enter Encryption Key (close with ^D): "));
+    fflush(stdout);
+  } else {
+    kfd = open(keyfile.c_str(), O_RDONLY);
+    if (kfd < 0) {
+      fprintf(stderr, _("Cannot open keyfile %s\n"), keyfile.c_str());
+      exit(1);
+    }
+  }
+  if (read(kfd, data, sizeof(data)) == 0) {
+    fprintf(stderr, _("Cannot read from keyfile %s\n"), keyfile.c_str());
+    exit(1);
+  }
+  if (kfd > 0) { close(kfd); }
+  StripTrailingJunk(data);
```

Comment:
This instruction is creating issues:
If you set a key `bscrypto -g /tmp/keyfile` and compare its content with the read `bscrypto -k /tmp/keyfile` the read is truncating the key.

`
keyfile=/tmp/keyfile
keyfile=/tmp/testkey
scratch:~ # bscrypto -d50 --dt -g $keyfile
21-Dec-2022 14:02:11.842632  (10): tools/bscrypto.cc:336-0 Generated passphrase = 5L|LVsO?8yusB~mB#m=ai5iz05UpZ/)=
21-Dec-2022 14:02:11.842846  (10): tools/bscrypto.cc:384-0 Keydata =
scratch:~ # bscrypto -d50 --dt -k $keyfile
21-Dec-2022 14:02:11.860430  (10): tools/bscrypto.cc:61-0 Key data = 5L|LVsO?
21-Dec-2022 14:02:11.860519  (10): tools/bscrypto.cc:462-0 Unwrapped passphrase = 5L|LVsO?
5L|LVsO?
scratch:~ # cat $keyfile
5L|LVsO?8yusB~mB#m=ai5iz05UpZ/)=
`

Reply:
While on previous version it was working
`bscrypto -d50 -g keyfile
 (40): lib/message.cc:230-0 Create daemon global message resource 1690bb0
 (10): tools/bscrypto.cc:310-0 Generated passphrase = $vOh3RL6WdP*oD8_Q9hp|mLBSHVaW&#&0
 (10): tools/bscrypto.cc:358-0 Keydata =
qt-kt:/tmp # bscrypto -d50 -k keyfile
 (40): lib/message.cc:230-0 Create daemon global message resource 1e0ebb0
 (10): tools/bscrypto.cc:398-0 Keydata = $vOh3RL6WdP*oD8_Q9hp|mLBSHVaW&#&
 (10): tools/bscrypto.cc:460-0 Unwrapped passphrase = $vOh3RL6WdP*oD8_Q9hp|mLBSHVaW&#
$vOh3RL6WdP*oD8_Q9hp|mLBSHVaW&#
`